src/tactic/congrm.lean

```diff
@@ -11,10 +11,83 @@ import tactic.interactive
 `congr` through the matching.  This allows more flexibility than `congr' n`, which enters uniformly
 through `n` iterations.  Instead, we can guide the matching deeper on some parts of the expression
 and stop earlier on other parts.
+
+##  Implementation notes
+
+###  Function underscores
+
+See the doc-string to `tactic.interactive.congrm` for more details.  Here we describe how to add
+more "function underscores".
+
+The pattern for generating a function underscore is to define a "generic" `n`-ary function, for some
+number `n`.  You can take a look at `tactic.congrm_fun_1, ..., tactic.congrm_fun_4`.
+These implement the "function underscores" `_₁, ..., _₄`.  If you want a different arity for your
+function, simply
+introduce
+```lean
+@[nolint unused_arguments]
+def congrm_fun_n {α₁ … αₙ ρ} {r : ρ} : α₁ → ⋯ → aₙ → ρ := λ _ … _, r
+notation `_ₙ` := congrm_fun_n
+```
+_Warning:_ `convert_to_explicit` checks that the first 18 characters in the name of `_ₙ` are
+identical to `tactic.congrm_fun_` to perform its job.  Thus, if you want to implement
+"function underscores" with different arity, either make sure that their names begin with
+`tactic.congrm_fun_` or you should change `convert_to_explicit` accordingly.
 -/
 
+/-- Get the list of explicit arguments of a function. -/
+meta def expr.list_explicit_args (f : expr) : tactic (list expr) :=
+tactic.fold_explicit_args f [] (λ ll e, return $ ll ++ [e])
+
 namespace tactic
 
+/--  Given a starting list `li`, a list `bo` of booleans and a replacement list `new`,
+read the items in `li` in succession and either replace them with the next element of `new` or
+not, using the list of booleans. -/
+def list.replace_if {α} : list α → list bool → list α → list α
```

Comment:
-> `data/list/defs.lean`

Reply:
Is this just `zip₃ cond`?

Reply:
Ah thanks, I thought it might be something like that